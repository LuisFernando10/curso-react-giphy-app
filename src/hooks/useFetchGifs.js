
    import { useEffect, useState } from "react"
    import { getGifs } from "../components/helpers/getGifs";

    export const useFetchGifs = ( category ) => {
      
        const [state, setState] = useState( {
            data: [],
            loadig: true
        } );

        useEffect( () => {

            getGifs( category )
                .then( imgs => {
                    setState({
                        data: imgs,
                        loadig: false 
                    });
                });
        }, [ category ] )

        return state;
    }
    