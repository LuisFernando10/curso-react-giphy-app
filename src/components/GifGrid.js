    
    import React from 'react';
    import PropTypes from 'prop-types';
    import { useFetchGifs } from '../hooks/useFetchGifs';
    import { GifGridItem } from './GifGridItem';

    export const GifGrid = ( { category } ) => {

        /*const [images, setImages] = useState( [] );

        /// El Hook 'UseEffect', permite ejecutar una única vez cuando el componente es renderizado por primera vez
        useEffect( () => {
            getGifs( category )
                .then( setImages );// Sólo se coloca la referencia al el método porque la promesa y el método reciben un parámetro, entonces se omite
        }, [ category ]);// Agregamos la 'categoría' para que si 'cambia', entonces ejecutamos nuevamente el Effect
        */

        const { data:images, loadig } = useFetchGifs( category ); // 'data:images', es la forma de renombrar una 'prop' cuando en el archivo local, queremos personalizar su nombre

        return (
            <>
                <h2>{ category }</h2>

                {/* { loadig ? <p>Cargando...</p> : null } //Ésta es la forma larga de evaluar dos condiciones  */}
                { loadig && <p className='animate__animated animate__shakeX'>Cargando...</p> }
                <div className='card-grid'>
                    {
                        images.map( ( img ) => (
                            <GifGridItem
                                key={ img.id }
                                { ...img }
                            />
                        ))
                    }
                </div>
            </>
        )
    }

    GifGrid.propTypes = {
        category: PropTypes.string.isRequired
    }