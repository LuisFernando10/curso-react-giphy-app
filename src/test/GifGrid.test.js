    
    import { GifGrid } from '../components/GifGrid';
    import { shallow } from 'enzyme';
    import { useFetchGifs } from '../hooks/useFetchGifs';
    
    // Forma de 'simular' o 'fingir' que ya traemos la información/data desde nuestro componente renderizado por nuesto CustomHook
    jest.mock('../hooks/useFetchGifs');

    describe('Pruebas en: GifGrid.js', () => { 
        
        const category = 'One Punch';

        test('Debe: Retornar componente < GifGrid />', () => { 

            // Aisgnamos el 'state' inicial del Hook y simulamos la data
            useFetchGifs.mockReturnValue({
                data: [],
                loading: true
            });

            const wrapper = shallow( <GifGrid category={ category }/> );
            expect( wrapper ).toMatchSnapshot();
         });

         test('Debe: Mostrar items cuando se cargan imágenes en useFetchGifs', () => { 
            
            const gifs = [
                {
                    id: 'ABC',
                    url: 'https://localhost/cualquier/cosa.png',
                    title: 'Cualquier Cosa'
                },
                {
                    id: 'DEF',
                    url: 'https://localhost/otra/cosa.png',
                    title: 'Otra Cosa'
                }
            ];
            
            useFetchGifs.mockReturnValue({
                data: gifs,
                loading: false
            });

            const wrapper = shallow( <GifGrid category={ category }/> );
            
            //expect( wrapper ).toMatchSnapshot();
            expect( wrapper.find('p').exists() ).toBe( false );
            expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );

          });
     })