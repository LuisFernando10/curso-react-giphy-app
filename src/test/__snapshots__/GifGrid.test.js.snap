// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en: GifGrid.js Debe: Retornar componente < GifGrid /> 1`] = `
<Fragment>
  <h2>
    One Punch
  </h2>
  <div
    className="card-grid"
  />
</Fragment>
`;
